source/slang/slang-ir-diff-jvp.cpp

```diff
@@ -221,102 +222,290 @@ struct DifferentiableTypeConformanceContext
+                if (auto genericBasePairStructType = as<IRStructType>(genericType))
+                {
+                    return as<IRFieldAddress>(builder->emitFieldAddress(
+                        builder->getPtrType((IRType*)
+                            findSpecializationForParam(
+                                ptrInnerSpecializedType,
+                                findField(ptrInnerSpecializedType, key)->getFieldType())),
                         baseInst,
-                        diffField->getKey()
+                        key
                     ));
+                }
+            }
+        }
+        else if (auto specializedType = as<IRSpecialize>(baseInst->getDataType()))
+        {
+            // TODO: Stopped here -> The type being emitted is incorrect. don't emit the generic's
+            // type, emit the specialization type.
+            // 
+            auto genericType = findInnerMostGenericReturnVal(as<IRGeneric>(specializedType->getBase()));
+            if (auto genericBasePairStructType = as<IRStructType>(genericType))
+            {
+                return as<IRFieldExtract>(builder->emitFieldExtract(
+                    (IRType*)findSpecializationForParam(
+                        specializedType,
+                        findField(genericBasePairStructType, key)->getFieldType()),
+                    baseInst,
+                    key
+                ));
+            }
+            else if (auto genericPtrType = as<IRPtrTypeBase>(genericType))
+            {
+                if (auto genericPairStructType = as<IRStructType>(genericPtrType->getValueType()))
+                {
+                    return as<IRFieldAddress>(builder->emitFieldAddress(
+                            builder->getPtrType((IRType*)
+                                findSpecializationForParam(
+                                    specializedType,
+                                    findField(genericPairStructType, key)->getFieldType())),
+                            baseInst,
+                            key
+                        ));
+                }
             }
         }
         else
         {
-            SLANG_UNREACHABLE("basePairType must be an IRStructType or PtrType<IRStructType>");
+            SLANG_UNEXPECTED("Unrecognized field. Cannot emit field accessor");
         }
         return nullptr;
     }
+
+    IRInst* emitPrimalFieldAccess(IRBuilder* builder, IRInst* baseInst)
+    {
+        return emitFieldAccessor(builder, baseInst, this->globalPrimalKey);
+    }
+
+    IRInst* emitDiffFieldAccess(IRBuilder* builder, IRInst* baseInst)
+    {
+        return emitFieldAccessor(builder, baseInst, this->globalDiffKey);
+    }
+
+    void relocateNewTypes(IRBuilder* builder)
+    {
+        for (auto typeInst : generatedTypeList)
+        {
+            moveGlobalToBeforeUses(builder, typeInst);
+        }
+    }
+
+    void _createGenericDiffPairType(IRBuilder* builder)
```

Comment:
If we get rid of generic type generation, we could just remove this, and remove the `moveGlobalToBeforeUses` logic as well.